lib/surface_site_web/live/app.ex

```diff
@@ -26,8 +27,11 @@ defmodule SurfaceSiteWeb.App do
               <div class="column is-4">
                 <CardWithIcon icon="fa-book" title="Documentation">
                   <p>
-                    The basic idea and motivation behind <b>Surface</b>, introducing the
-                    all the main features along with <strong>live examples</strong> and detailed explanation to
+                    The basic idea and motivation behind
+                    <b>Surface</b>, introducing the
+                    all the main features along with
+                    <strong>live examples</strong>
+                    and detailed explanation to
                     help you to get started.
                   </p>
```

Comment:
This is another instance where the `siblings go on separate lines` rule and `retain existing lack-of-whitespace` rules collide to make an odd result.

But it also highlights that the formatter sees a `\n` in a text node and splits it into separate text nodes.

How do people feel about this? Would it be preferable for the formatter to "move around" newlines in a text node to meet/maintain the target line length?

E.g.

```html
The basic idea and motivation behind <b>Surface</b>, introducing the
all the main features along with <strong>live examples</strong> and detailed explanation to
help you to get started.
```

might become 

```html
The basic idea and motivation behind <b>Surface</b>, introducing the all the main features along
with <strong>live examples</strong> and detailed explanation to help you to get started.
```

The formatter would detect that the `<b>` and `<strong>` nodes were placed right inside of the text, not on their own line.

Reply:
> How do people feel about this? Would it be preferable for the formatter to "move around" newlines in a text node to meet/maintain the target line length?

I guess so.

Reply:
An alternative would be, the formatter just leaves the line alone unless it breaks the line length limit. I think this behavior would keep things more consistent with the Elixir formatter.

Any thoughts?

Reply:
@msaraiva In light of the above discussion, how would you prefer the formatter to treat these examples?

```html
Foo  bar   baz <div>   <b> Surface     </b> </div>
```

```html
<section class="great-gpa"> <div class="gpa"> <div class="parent"> <p>Content</p> </div> </div> </div>
```

Reply:
Hi @paulstatezny! 

I believe the best would be to follow the "try to meet/maintain the target line length" rule. So I guess, it would be something like:

```
Foo bar baz<div><b>Surface</b></div>
```

and

```
<section class="great-gpa">
  <div class="gpa"><div class="parent"><p>Content</p></div></div> 
</section>
```

Assuming the target line length is `100` here, it would be enough to just put `</section>` on a new line but I believe if we move the closing tag, we also should move the opening tag. Right?

Another thing that comes to my mind is if we should try to preserve user's new lines before tags that are not "inline" by default like `<div>` and `<p>`. It seems to me that if the user writes the following, we should keep it:

```
This is a text:
<p>The text</p>
```

Another way to think about that would be to always preserve user's new lines unless we already have to "touch" that line due to its size. That means we would only combine or split lines that either:

  - Is larger than max_line_length
  - Have been already modified by the formatter.

I'm not sure what's the best way to implement that and I confess I don't have much knowledge on the subject when it comes to the best way to implement those formatting rules. I'm also not sure if having this kind of conditions forcing the formatter to try multiple possibilities might affect the performance badly or not. Have you seen how other similar formatters do this stuff?

Reply:
@msaraiva Thanks for your thoughts!

I could see deciding what to move based on line length.

> but I believe if we move the closing tag, we also should move the opening tag. Right?

And agreed there. 👍

> should [we] try to preserve user's new lines before tags that are not "inline" by default like `<div>` and `<p>`.

So, some formatters (like prettier, which is one of the best HTML ones out there as far as I can tell) make this distinction and behave differently if you use `div` instead of `span`. That's tricky for surface, because what about Surface components that are intended to be used the same way? E.g.

```HTML
<p>
  Some docs referencing <InlineCode>variable</InlineCode>
</p>
```

(Imagine `InlineCode` is a component that wraps it in a span with some Tailwind classes applied.)

We can keep a list of inline or block tags, but I think will show itself as "not smart enough" in cases like these.

That's an interesting idea of giving the formatter more "freedom" to modify lines we've already touched. I'll have to think about that more.

> I'm also not sure if having this kind of conditions forcing the formatter to try multiple possibilities might affect the performance badly or not.

We're already doing this when it comes to wrapping opening tags. (Building out the 1-line version, checking its length, and then re-building on multiple lines.) It still runs quite fast, although adding more of these might start bogging it down. We can always benchmark.

Reply:
@msaraiva And if we don't maintain a list of inline/block elements, then it's difficult to know to leave this alone:

```html
Text prose <b>bold</b> words
```

While editing this:

```html
Some text <div>other text</div> more text
```

However, I tried to think outside the box and yesterday I came up with a different approach: if the user has in line elements among text, and those elements I don't have other elements inside them, leave them inline.￼￼￼ Otherwise move them to their own line.

After implementing this rule, I ran it on the surface site codebase and all of those problematic areas stopped getting edited by the formatter.

I think this is a good way to proceed forward for now, with the least amount of assumptions and complexity added.

I will release a new version and put in a separate PR, if that works for you.￼￼￼￼￼￼

Reply:
@paulstatezny sounds good to me! I believe identifying inline elements and handling them differently will eventually be useful. Regarding components that we want to treat as inline, since they are not supposed to be a lot, we could have an option for `.formatter.exs` to list them.

Reply:
That's a great idea. I can see it being helpful treating different elements differently. I wanted to buy more time to think about it and weigh possibilities. 😁

I did some refactoring today that should make it easier to customize how we modify whitespace in a number of scenarios.